prompt2model/dataset_transformer/prompt_based.py

```diff
@@ -0,0 +1,122 @@
+"""A simple dataset transformer that uses a plan prompt and transform prompt."""
+from __future__ import annotations
+
+import asyncio
+from collections.abc import Callable
+
+import datasets
+
+from prompt2model.dataset_transformer.base import DatasetTransformer
+from prompt2model.dataset_transformer.prompt_template import (
+    construct_prompt_for_plan,
+    construct_prompt_for_transform_data,
+)
+from prompt2model.prompt_parser import PromptSpec
+from prompt2model.utils import (
+    API_ERRORS,
+    api_tools,
+    get_formatted_logger,
+    handle_api_error,
+)
+from prompt2model.utils.parse_responses import make_single_api_request, parse_json
+
+logger = get_formatted_logger("DatasetTransformer")
+
+
+class PromptBasedDatasetTransformer(DatasetTransformer):
+    """Transform data based on a transform prompt."""
+
+    def __init__(
+        self,
+        plan_prompt_fn: Callable[
+            [str, list[dict], str], str
+        ] = construct_prompt_for_plan,
+        transform_prompt_fn: Callable[
+            [str, dict, str, str], str
+        ] = construct_prompt_for_transform_data,
+    ):
+        """Initialize the class."""
+        self.plan_prompt_fn = plan_prompt_fn
+        self.transform_prompt_fn = transform_prompt_fn
+        self.plan: str = ""
+
+    def make_dataset_from_samples(
+        self,
+        inputs: list[str],
+        outputs: list[str],
+    ) -> datasets.DatasetDict:
+        """Given a list of inputs and outputs, make a dataset."""
+        if len(inputs) <= 0 or len(inputs) != len(outputs):
+            raise ValueError("Length of inputs and outputs must be >0 and equal.")
+
+        dataset_dict = {}
+        dataset_dict["train"] = datasets.Dataset.from_dict(
+            {"input_col": inputs, "output_col": outputs}
+        )
+        return datasets.DatasetDict(dataset_dict)
+
+    def transform_data(
+        self,
+        prompt_spec: PromptSpec,
+        dataset: datasets.Dataset,
+        num_points_to_transform: int,
+    ) -> datasets.DatasetDict:
+        """Transform the dataset according to the prompt_spec and dataset."""
+        plan_prompt = self.plan_prompt_fn(
+            prompt_spec.instruction,
+            dataset,
+            prompt_spec.examples,
+        )
+        self.plan = make_single_api_request(plan_prompt)
+
+        logger.info(f"Plan created. Plan: {self.plan}")
+
+        inputs = []
+        outputs = []
+
+        required_keys = ["input", "output"]
```

Comment:
nit - move this closer to the place it is used. I think it's also fine to just pass this list as a literal to the function rather than passing in the variable, for better readability